### DISCUSSION TITLE
Implementing indirect instruction execution in processor plugin

### DISCUSSION THREAD

**Discussion by shuffle2** [NationalSecurityAgency/ghidra#6612] [Q&A] [answered]
The Andes V3 instruction set has an extension named EX9.IT.
EX9.IT extension adds a system register ITB (Instruction_Table Base) which points to a memory location which will be treated as a 512-entry array of 32bit values.
The EX9.IT instruction takes an immediate value which is used to index into this table.
32bits from the given index are fetched, and a single instruction is executed (if the 32bit data begins with a 16bit instruction, only 1 instruction is executed).
The feature is designed to allow executing a 32bit instruction from a 16bit one (EX9.IT is 16bit insn), the idea being that the most frequently used 32bit instructions can be extracted from the main executable and put into Instruction_Table, resulting in overall binary size decrease.

Usage looks something like:

```
# set ITB=0x4040
MOVI      a0, 0x4040
MTUSR     a0, ITB
...
# execute 1 instruction from Instruction_Table[0x10] inline
EX9.IT    0x10
...
```

Enlightening a disassembler about this has 2 annoyances:
1. ITB register value must be tracked
2. Upon EX9.IT disassembly, parser must be redirected to handle value fetched from ITB, while still considering the resulting instruction as "inline". 

1 seems a common problem, and can be handled in ghidra similar to tracking other global pointers.
2 Seems more tricky, I'm not sure how to tackle this in sleigh. What is a way to handle this?

**Comment by shuffle2** [comment]
I realized this construct is sort of like an inverted delayslot. However, I don't think ghidra's existing delayslot behavior can be abused to implement it.

**Comment by shuffle2** [comment]
I implemented it with callotherfixup dynamic pcode and it seems to work well enough (although would be nicer if didn't have to do it this way)

**Comment by GhidorahRex** [comment]
The only way I can think of would be having a hidden return address register and context value that flags an indirect execution:
```
register ... [ returnadd ];
indir=(0,0)
...
:EX9.IT ... [indir=1;] {
    addr = ITB + offset;
    returnadd = inst_next;
    call [addr];
}
```

and for every instruction that can potentially be used in the table, have a return check:
```
rtn_check: indir=0 {}
rtn_check:indir=1[indir=0;]{ return [returnadd];}
...
:add regA, regB is opcode=0x1 & regA & regB & rtn_check {
    regA = regA + regB;
    build rtn_check;
}
```

I haven't tested this, but regardless this will only work if the instructions used in the `EXT9.IT` instruction are used as one-off instructions. It definitely will *not* work if they are instructions that could be used in a non-inline context and an inline context.

**Comment by shuffle2** [comment]
Thanks for the suggestion. However, the instructions referenced by `EX9.IT` are used in both non-inline and inline contexts.

My pcode injection looks like this: (basically, using PseudoDisassembler to get the pcode of the referenced instruction)
```java
public class InjectEX9IT extends InjectPayloadCallother {
	private PcodeOp[] EMPTY_PCODEOP = new PcodeOp[0];
	private int INSTRUCTION_TABLE_ENTRY_LENGTH = 4;
	private AddressSpace defaultSpace;
	private CodeUnitFormat codeUnitFormat;
	private Register itbReg;
	private Register contextReg;

	public InjectEX9IT(String sourceName, SleighLanguage language) {
		super(sourceName);
		itbReg = language.getRegister("ITB");
		contextReg = language.getRegister("contextreg");
		defaultSpace = language.getAddressFactory().getDefaultAddressSpace();
		codeUnitFormat = new CodeUnitFormat(new CodeUnitFormatOptions());
	}

	PseudoInstruction disasmAt(Program program, Address disasmAddr, Address fetchAddr) {
		BigInteger context = program.getProgramContext().getValue(contextReg, disasmAddr, false);
		if (context == null || context == BigInteger.valueOf(0)) {
			throw new IllegalArgumentException("ITMode 0 in EX9.IT");
		}
		PseudoDisassembler disassembler = new PseudoDisassembler(program);
		byte[] data = new byte[INSTRUCTION_TABLE_ENTRY_LENGTH];
		try {
			if (program.getMemory().getBytes(fetchAddr, data) != Array.getLength(data)) {
				return null;
			}
			return disassembler.disassemble(disasmAddr, data);
		} catch (Exception ex) {
			return null;
		}
	}

	@Override
	public PcodeOp[] getPcode(Program program, InjectContext con) {
		Address ex9itAddr = con.baseAddr;
		int imm9u = (int) con.inputlist.get(0).getOffset();
		int itOffset = imm9u * INSTRUCTION_TABLE_ENTRY_LENGTH;

		BigInteger ITB = program.getProgramContext().getValue(itbReg, ex9itAddr, false);
		if (ITB == null) {
			return EMPTY_PCODEOP;
		}
		long memOffset = (ITB.longValue() & ~0b11) + itOffset;

		Address fetchAddr = defaultSpace.getAddress(memOffset);
		PseudoInstruction insn = disasmAt(program, ex9itAddr, fetchAddr);
		// Could be bad ITB
		if (insn == null) {
			return EMPTY_PCODEOP;
		}

		// Set comment if there's a valid insn referenced.
		// TODO append the referenced instruction in a more disassembler-aware way
		Listing listing = program.getListing();
		if (listing.getComment(CodeUnit.EOL_COMMENT, ex9itAddr) == null) {
			// getRepresentationString is also slow
			String ex9itComment = codeUnitFormat.getRepresentationString(insn);
			program.withTransaction("set EX9.IT comment", () -> {
				listing.setComment(ex9itAddr, CodeUnit.EOL_COMMENT,
						String.format("%s {%s}", fetchAddr.toString(), ex9itComment));
			});
		}

		String mnem = insn.getMnemonicString();
		if (mnem == "EX9.IT") {
			// hw would generate Reserved Instruction Exception
			return EMPTY_PCODEOP;
		}

		// 32bit insns which use inst_next (PC + 4) need to be fixed up to use PC + 2,
		// since EX9.IT is 16bit. J/JAL also OR displacement instead of ADDing it.
		// if J : PC = concat(PC[31,25], (Inst[23,0] << 1)) // not signed? and OR instead of ADD
		// if JAL : R30 = PC + 2; PC = concat(PC[31,25], (Inst[23,0] << 1))
		// JRAL, JRAL.xTON, JRALNEZ, BGEZAL, BLTZAL: RT = PC + 2
		// These are taken care of in sleigh code.
		//String[] pcRel = { "J", "JAL", "JRAL", "JRAL.xTON", "JRALNEZ", "BGEZAL", "BLTZAL" };
		//if (Arrays.asList(pcRel).contains(mnem)) {
		//	return EMPTY_PCODEOP;
		//}

		return insn.getPcode();
	}

}
```

This is working well except for the issue of 32bit instructions which would normally use `inst_next` in their semantic code.

I had thought I could do something like this:
```
define register offset=0x110 size=4 [bi_addr contextreg];

define context contextreg
    ITMode = (0,0)
;

inst_next_it: addr is ITMode=0 [addr = inst_start + 4;] { export *:4 addr; }
inst_next_it: addr is ITMode=1 [addr = inst_start + 2;] { export *:4 addr; }

imm24s_rel: rel is s0_23 [ rel = inst_start + (s0_23 << 1); ] { export *:4 rel; }
:JAL imm24s_rel is u24_24=1 & imm24s_rel & inst_next_it & ITMode=0 {
    lp = inst_next_it;
    call imm24s_rel;
}
jal_dst_it: rel is s0_23 [ rel = (inst_start & 0xfe000000) | (s0_23 << 1); ] { export *:4 rel; }
:JAL jal_dst_it is u24_24=1 & jal_dst_it & inst_next_it & ITMode=1 {
    lp = inst_next_it;
    # XXX call pcodeop implicitly uses inst_next which generates improper code xref at +4
    call jal_dst_it;
}

:EX9.IT imm9u is op9_14=0b110101 & imm9u     [ ITMode=1; ] { ex9it(imm9u:2); }
:EX9.IT imm5u is op5_14=0b1011101010 & imm5u [ ITMode=1; ] { ex9it(imm5u:1); }
```

However, the `contextreg` doesn't seem to propagate to the program context state which gets accessed by the injection java code (the result of `program.getProgramContext().getValue(contextReg, disasmAddr, false)` is `null`).

The discrepancy in behavior is outlined here:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/113063/139d1086-cc8b-46f7-b86c-25c144b75168)

Is this a good approach? It seems _almost_ working how I'd like it (at least decompiler receives proper pcode). There's just a few instructions which need to operate differently in case they're executed via `EX9.IT`. I had also considered generating the pcode for those instructions in java, but was hoping to keep it in sleigh if possible.

**Comment by shuffle2** [comment, marked as answer]
I eventually arrived at something that seems to produce good results, although I think it may highlight some issues in existing ghidra code.

I split the functionality for decoding the indirectly referenced instructions, then _only_ return pcode in the injection hook, and perform the extra annotation in an Analyzer:
```java
package plugin.core.analysis;
//...
public class EX9ITDisassembler {
	private Program program;
	private ProgramContext programContext;
	private Language language;
	private Memory memory;
	private final int INSTRUCTION_TABLE_ENTRY_LENGTH = 4;
	private Register itbReg;
	private Listing listing;
	private Address zeroAddress;
	private final String EX9IT_MNEMONIC = "EX9.IT";

	public EX9ITDisassembler(Program program) {
		this.program = program;
		memory = program.getMemory();
		language = program.getLanguage();
		programContext = program.getProgramContext();
		listing = program.getListing();
		itbReg = language.getRegister("ITB");
		zeroAddress = program.getAddressFactory().getDefaultAddressSpace().getAddress(0);
	}

	boolean instructionIsEX9IT(Instruction insn) {
		return insn.getMnemonicString().equals(EX9IT_MNEMONIC);
	}

	public PseudoInstruction disassemble(Address disasmAddr, byte bytes[]) throws InsufficientBytesException,
			UnknownInstructionException {
		PseudoDisassemblerContext disassemblerContext = new PseudoDisassemblerContext(programContext);
		MemBuffer memBuffer = new ByteMemBufferImpl(disasmAddr, bytes, language.isBigEndian());

		// check that address is defined in memory
		try {
			memBuffer.getByte(0);
		} catch (Exception e) {
			return null;
		}

		InstructionPrototype prototype = null;
		disassemblerContext.flowStart(disasmAddr);
		prototype = language.parse(memBuffer, disassemblerContext, false);
		if (prototype == null) {
			return null;
		}

		PseudoInstruction instr;
		try {
			// First, normal decode
			instr = new PseudoInstruction(program, disasmAddr, prototype, memBuffer, disassemblerContext);

			// hw would generate Reserved Instruction Exception
			if (instructionIsEX9IT(instr)) {
				return null;
			}

			// If it's branch, it's decoded as if current pc is 0
			// Must avoid passing program to PseudoInstruction, otherwise it will read from
			// program at the given addr - which we're trying to avoid
			FlowType flowType = prototype.getFlowType(instr);
			if (flowType.isCall() || flowType.isJump()) {
				instr = new PseudoInstruction(program.getAddressFactory(), zeroAddress, prototype, memBuffer,
						disassemblerContext);
			}
		} catch (AddressOverflowException e) {
			throw new InsufficientBytesException(
					"failed to build pseudo instruction at " + disasmAddr + ": " + e.getMessage());
		}

		return instr;
	}

	Instruction getEX9ITInstruction(Address address) {
		Instruction insn = listing.getInstructionAt(address);
		if (insn == null) {
			return null;
		}
		if (!instructionIsEX9IT(insn)) {
			return null;
		}
		return insn;
	}

	public PseudoInstruction getITInstruction(Instruction ex9itInsn) {
		Address ex9itAddress = ex9itInsn.getAddress();
		long itOffset = ex9itInsn.getScalar(0).getUnsignedValue() * INSTRUCTION_TABLE_ENTRY_LENGTH;

		BigInteger ITB = programContext.getValue(itbReg, ex9itAddress, false);
		if (ITB == null) {
			return null;
		}
		long memOffset = (ITB.longValue() & ~0b11) + itOffset;
		Address fetchAddress = ex9itInsn.getAddress().getNewAddress(memOffset);

		byte[] data = new byte[INSTRUCTION_TABLE_ENTRY_LENGTH];
		try {
			if (memory.getBytes(fetchAddress, data) != Array.getLength(data)) {
				return null;
			}
			return disassemble(ex9itAddress, data);
		} catch (Exception ex) {
			return null;
		}
	}

	public PseudoInstruction getITInstruction(Address ex9itAddress) {
		Instruction ex9Insn = getEX9ITInstruction(ex9itAddress);
		if (ex9Insn == null) {
			return null;
		}
		return getITInstruction(ex9Insn);
	}
}
```

```java
package ghidra.app.util.pcodeInject;
//...
public class InjectEX9IT extends InjectPayloadCallother {

	public InjectEX9IT(String sourceName) {
		super(sourceName);
	}

	@Override
	public PcodeOp[] getPcode(Program program, InjectContext con) {
		EX9ITDisassembler disassembler = new EX9ITDisassembler(program);
		PseudoInstruction itInsn = disassembler.getITInstruction(con.baseAddr);
		// Could be bad ITB
		if (itInsn == null) {
			return null;
		}

		// NOTE SymbolicPropogator must be patched to allow STORE pcode ops
		return itInsn.getPcode();
	}

}
```

```java
package plugin.core.analysis;
//...
public class AndeStarEX9ITAnalyzer extends AbstractAnalyzer {
    private final static String PROCESSOR_NAME = "AndeStar";
    private final static String NAME = "AndeStar EX9IT Analyzer";
    private final static String DESCRIPTION = "Annotates EX9.IT instructions";

    private final static CodeUnitFormat codeUnitFormat = new CodeUnitFormat(new CodeUnitFormatOptions());

    public AndeStarEX9ITAnalyzer() {
        super(NAME, DESCRIPTION, AnalyzerType.INSTRUCTION_ANALYZER);
        setDefaultEnablement(true);
    }

    @Override
    public boolean canAnalyze(Program program) {
        return program.getLanguage().getProcessor().equals(
                Processor.findOrPossiblyCreateProcessor(PROCESSOR_NAME));
    }

    @Override
    public boolean added(Program program, AddressSetView set, TaskMonitor monitor, MessageLog log)
            throws CancelledException {
        Listing listing = program.getListing();
        ReferenceManager refMgr = program.getReferenceManager();

        EX9ITDisassembler disassembler = new EX9ITDisassembler(program);

        for (Address addr : set.getAddresses(true)) {
            PseudoInstruction itInsn = disassembler.getITInstruction(addr);
            if (itInsn == null) {
                continue;
            }

            // Add a comment
            // TODO append to mnemonic instead?

            // itInsn will not have associated program if it's a branch
            if (itInsn.getProgram() != null) {
                String comment = codeUnitFormat.getRepresentationString(itInsn);
                listing.setComment(addr, CodeUnit.EOL_COMMENT, comment);
            } else {
                // dont really need extra comment - ghidra will add one because of the reference
                // comment = itInsn.getPrimaryReference(0).getToAddress().toString();
            }

            // Copy the references

            refMgr.removeAllReferencesFrom(addr);

            Reference[] refs = itInsn.getReferencesFrom();
            if (refs.length > 0) {
                CodeUnit cu = listing.getCodeUnitAt(addr);
                for (Reference ref : refs) {
                    cu.addMnemonicReference(ref.getToAddress(), ref.getReferenceType(), ref.getSource());
                }
            }
        }

        return true;
    }
}
```

The things to highlight:
* `PseudoInstruction` is apparently intended to allow my exact use case: disassembling an instruction as-if it exists at a given `Address`, while supplying arbitrary `byte[]` data to decode. However, it doesn't fully work. To actually use `PseudoInstruction` like this, you must not construct it with a `Program` reference - a path which isn't exposed by `PseudoDisassembler`. Additionally, `PseudoInstruction`s created like this will still try to read data from `program` when e.g. `getRepresentationString` is called (instead of reading from the provided `byte[]` data). So, you either wind up with null deref exception or incorrect disasm output.
* I am patching [some code in ghidra ](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/src/main/java/ghidra/program/util/SymbolicPropogator.java#L857C7-L859C8)to allow injected pcode STORE ops to be processed. I'm not sure why they are unconditionally ignored by existing ghidra code.
* I really wish injected pcode were still visible in the gui (disasm view), like other pcode.

**Comment by emteere** [comment]
@shuffle2 I was thinking about another way to make this work, or an extension we might add to sleigh parsing to handle something like this.

There are two other ways this could work, one is that the instruction bytes the EXEC.IT instruction uses are fetched in an analyzer and then set in a 4 byte context field.  To make this work, all instructions would need to be parsed from the context and not from memory.  It might get complicated with variable sized instructions.  But they can actually consume bytes as well.  This would be possible with the addition of ":^instruction" type context loading, in conjunction with an analyzer that would set the context correctly and then re-parse the instruction.  This would work, but unfortunately all RISC-V instructions would need to be parsed from context.  So the tokens that are used to parse from memory would be moved into tokens in the context register.

The other would require a change to sleigh that could load/append bytes to the parse buffer after the IT instruction, most likely in the [ action ] part of the parsing.  And then re-curse the parsing with the again.  Something like:

[ inst_buffer = read(space, target, 4); ]  This would append the bytes, or replace the bytes in the parse buffer read from the target location, which could also be from the instruction Hardware lookup table implemented as another address space.  This might also work for peeking at bytes and matching during the parse without consuming them.

I'm not sure this is the best solution either.  We'd need to think about the changes that would be necessary to the sleigh parsing.

Also, are you considering submitting a PR, or is the spec available somewhere to take a look?

**Reply by shuffle2** [reply]
Hi @emteere , i see you've implemented a similar feature (`nds.exec.it`) for andes v5: https://github.com/NationalSecurityAgency/ghidra/commit/ade1ad71ba5b80e910fcd98ad688f658439724a5

nice! does it work as expected in the emulator, decompiler, etc? I will need to try it out.

I have my nds32 spec here: https://github.com/shuffle2/ghidra-AndeStar , please take a look. would still be interested for the nds32 support in ghidra to be improved / for ex9.it and ifcall to be properly supported.

**Comment by emteere** [comment]
Hoping someone can supply a good sample example, or try it out locally on a non-shareable sample that uses the instruction both for v5 and v3, and give us feedback on the implementation.

The intent was to handle both the table definition defined somewhere in memory, and the random instruction address referenced directly by the instruction.  It is a placeholder implementation, which in "theory" will work.

I believe it should work in the decompiler and emulator if I understand correctly how the implementation would appear in the wild.  The listing/graph/basicblock handling and actual function body definitions may need some careful care as the instruction would be shared by many functions, and would have to be protected in certain analyzers.

We've been reviewing the an NDS32 PR and once verified that this is the right way to go, implement the ex9.it instruction as well for the NDS32.

**Reply by shuffle2** [reply]
Here is hopefully a good representative binary for nds32 EX9.IT usage. IFCALL is also used.
https://drive.google.com/file/d/1UVWyUKu1D6NGZWaXQvNKT_W669DX3j58/view?usp=sharing
It's in ghidra zip format - i haven't shared things that way before, lmk if you want a different format. I'm assuming you can either use my nds32 plugin, or extract the binary and load it yourself.

This is interesting example, because the ITB register has 2 valid values: 0x25964, which is used initially, assigned here:
```
        000241e8 44  02  59                      MOVI       a0 ,0x25964
                 64
        000241ec 42  0e  00                      MTUSR      a0 ,ITB
                 21
```
and then 0xdc208, assigned here:
```
        000df460 50  00  02       030           ADDI       a0 ,a0 ,0x208
                 08
        000df464 42  0e  00       030           MTUSR      a0 ,ITB
                 21
```
this relates to rom context, and then the "rom patch" context, which overrides some functionality of rom, but does not entirely replace it.
You can imagine, someone might want to have a ghidra database dedicated to the rom, and then other databases which include the rom and specific patch versions.

p.s. i added the ex9 emu patch i was using here: https://github.com/shuffle2/ghidra-AndeStar/blob/main/ex9_emu.patch

**Reply by shuffle2** [reply]
Here is a binary for v5: https://drive.google.com/file/d/17NhqkYQUzrrJLUlR5fRZYCJVDx2d0IXS/view?usp=sharing , altho it's missing the Instruction_Table for the rom context (the IT for fw context *is* present). I've gotta figure out how to dump the rom's IT, yet :) In any case, it should be enough to see how it's working (check references to `uitb`). Note that I haven't properly setup the ExecTable here...

I would note that "ex9" mnemonic appears deprecated in v5 parlance, see https://github.com/andestech/andes-v5-isa/releases (they're now `nds.exec.it` and `nds.nexec.it`).

some v5 csr encodings are here: https://github.com/andestech/nds-openocd/blob/fcb3417d2e0e0c63a2d7ead2737f1ccc0b201f20/src/target/riscv/ndsv5_encoding.h#L203 (there are quite a lot in addition to UITB)

in both v3/v5, PC for JAL instructions needs special care: https://github.com/andestech/qemu/blob/922ee62635859afccca41a9cee7361639d7f4c3c/target/riscv/insn_trans/trans_xandescodenseops.c.inc#L96 (V3 is a bit different).

It seems like e.g. "c0 91", "a0 91", "a8 92" etc should decode to `nds.nexec.it`, and currently does not.
* edit: realized this is because i must set ecdv=1 in contextreg.  maybe that should be the default tho? 🤷‍♂️ Apparently this indicates "version" of the CoDense hardware. The version field is 2 bits wide, and current value appears to be `0b01`.

Another thing I've noticed is that ghidra decodes some nds instructions as if they're "rv32k" instructions, e.g. `NDS.LWGP` is decoded as `saes32.encsm`. Commenting out the inclusion of the "rvk32k.sinc" file fixes this. I wonder if there are other conflicts occuring?

**Reply by shuffle2** [reply]
also for v5, the decoding of nds.lwgp is incorrect. this change is needed:
```diff
diff --git a/Ghidra/Processors/RISCV/data/languages/andestar_v5.instr.sinc b/Ghidra/Processors/RISCV/data/languages/andestar_v5.instr.sinc
index bb30a6333a..b8c6caad7a 100644
--- a/Ghidra/Processors/RISCV/data/languages/andestar_v5.instr.sinc
+++ b/Ghidra/Processors/RISCV/data/languages/andestar_v5.instr.sinc
@@ -20,7 +20,7 @@ simm18_lh: val is sop3131 & op1516 & op1719 & op2020 & op2130 [ val = (sop3131<<
        export *[const]:$(XLEN) val;
 }

-simm18_lw: val is sop3131 & op1516 & op1719 & op2020 & op2130 [ val = (sop3131<<18) | (op1516<<16) | (op1719<<13) | (op2020<<12) | (op2130<<2); ] {
+simm18_lw: val is sop3131 & op2121 & op1516 & op1719 & op2020 & op2230 [ val = (sop3131<<18) | (op2121 << 17) | (op1516<<15) | (op1719<<12) | (op2020<<11) | (op2230<<2); ] {
        export *[const]:$(XLEN) val;
 }
 ```
 
 Will have to go through the other specs to double check them.

**Reply by emteere** [reply]
Thanks, will take a look!

**Reply by emteere** [reply]
I know there have been some internal changes to the NDS32 to get the PR to acceptance.  There should be a few more and then it will get pushed to the REPO.  Hopefully to make it into 12.0 release.  I'll see if I can factor in your changes to the v3/v5, if not then 12.0.1 release.

**Reply by shuffle2** [reply]
Another issue with v5 i've noticed is that [Zcmp/Zcmt instructions](https://github.com/andestech/qemu/blob/922ee62635859afccca41a9cee7361639d7f4c3c/target/riscv/insn16.decode#L200C23-L200C32) aren't supported by ghidra, so a bunch of things get decoded as `c.fsdsp`  incorrectly.

edit: actually, there is use of instruction like:
101 01000... 10
which doesn't seem to match any Zc* instructions...maybe they've implemented an old version of the spec which encoded this pattern?
edit2: ...on closer inspection, the rest of the Zc ops don't really seem to fit either, when considering surrounding code. might have to spend some time experimenting to figure out what the instructions are actually doing on hardware. I'd suspect it's some core like Andes V5 N225, which is advertised to support Zc...

**Reply by shuffle2** [reply]
Considering that value of `uitb` in rom context is zero (the reset value), and I haven't been able to find anything looking like an instruction table in the address space accessible to rom, I wonder if encoding of `nds.nexec.it` on this core does not follow the documentation (and maybe, `ECDV==0`, I haven't checked that yet...)

It would also make sense if the "c.fsdsp" encoding space was actually used for "nds.exec.it" - since they only appear in the firmware code, after the firmware has setup `uitb` to nonzero value.
Sorry for not knowing this ahead of time :')

I'll have to collect some more info and report back.

**Reply by shuffle2** [reply]
Sorry for repeated replies, but yea, it seems like the v5 example is "PP16", and not CoDense.
Here are some csr values from the core:
```
csr name          value
f11 mvendorid        31e     (ANDES_VENDOR_ID)
f12 marchid     80000045     (ANDES_CPUID_N45)
f13 mimpid           640
f14 mhartid            0
301 misa        40941105     ( A, C, I, M, S, U, X )
7c0 milmb         800001
7c1 mdlmb        2200001
800 uitb               0
fc0 micm_cfg       61319
fc1 mdcm_cfg       59319
fc2 mmsc_cfg    c005e028
fc3 mmsc_cfg2         40
```
so it has `mmsc_cfg.ECD=1, mmsc_cfg2.{ECDV=0,PP16=1}`, meaning neither of the documented forms of `nds.{n}exec.it` apply, and it's probably PP16, whatever that is 🙃 sorry for the confusion!

for ghidra v5 support, i guess it should also be aware of pp16 in the context (to disable CoDense), however afaik we need to reverse the pp16 instruction encodings - don't see it documented.

(my note about nds.lwgp should still apply, as that is part of "andes performance extensions", not CoDense)

as `uitb` _is_ set by firmware, and there is clear instruction table present, it does seem like PP16 contains similar ex9.it-like functionality, just encoded differently.

**Comment by shuffle2** [comment]
@emteere here are changes (verified against hardware) that implement "pp16"-style exec.it and push/pop: (hah, maybe "pp16" is just for "16bit Push/Pop"...): https://gist.github.com/shuffle2/87a4a661e49b1a0f65b1132a9c6d1c92

push/pop seems to work ok. disassembly listing of the reglist could be nicer.

exec.it doesn't work correctly, although maybe it's because i'm not using it correctly? I created a memory region of ExecTable address space and copied the instruction table data there. ghidra creates references from the exec.it instructions to the correct place in ExecTable. However,  the disassembler does not follow flow correctly, and neither does the decompiler (get error `Low-level Error: Could not find op at target address: (ExecTable,0x00dc)`).
There is the additional complication with `JAL` executed via exec.it, which I mentioned before (but maybe that can just be accounted for in sleigh for `JAL`?).
